cmd/whitesourceExecuteScan.go

```diff
@@ -90,89 +126,183 @@ func newWhitesourceUtils() *whitesourceUtilsBundle {
 	log.Entry().Info("-----------------------------------------------------")
 
-	if config.Reporting || config.SecurityVulnerabilities {
-		// Project was scanned. We need to wait for WhiteSource backend to propagate the changes
-		// before downloading any reports or check security vulnerabilities.
-		if err := pollProjectStatus(config, sys); err != nil {
-			return err
-		}
+	if err := checkAndReportScanResults(config, scan, utils, sys); err != nil {
+		return err
 	}
 
+	if err := persistScannedProjects(config, scan, utils); err != nil {
+		return fmt.Errorf("failed to persist scanned WhiteSource project names: %w", err)
+	}
+
+	return nil
+}
+
+func checkAndReportScanResults(config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils, sys whitesource) error {
+	if !config.Reporting && !config.SecurityVulnerabilities {
+		return nil
+	}
+	if err := blockUntilReportsAreaReady(config, scan, sys); err != nil {
+		return err
+	}
 	if config.Reporting {
-		paths, err := downloadReports(config, utils, sys)
+		paths, err := downloadReports(config, scan, utils, sys)
 		if err != nil {
 			return err
 		}
 		piperutils.PersistReportsAndLinks("whitesourceExecuteScan", "", nil, paths)
 	}
-
 	if config.SecurityVulnerabilities {
-		// Check for security vulnerabilities and fail the build if cvssSeverityLimit threshold is crossed
-		if err := checkSecurityViolations(config, sys); err != nil {
+		if err := checkSecurityViolations(config, scan, sys); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func resolveProjectIdentifiers(config *ScanOptions, utils whitesourceUtils, sys whitesource) error {
-	if config.ProjectName == "" || config.ProductVersion == "" {
-		coordinates, err := utils.GetArtifactCoordinates(config)
+func resolveProjectIdentifiers(config *ScanOptions, scan *whitesourceScan, utils whitesourceUtils, sys whitesource) error {
+	if scan.aggregateProjectName == "" || config.ProductVersion == "" {
+		options := &versioning.Options{
+			ProjectSettingsFile: config.ProjectSettingsFile,
+			GlobalSettingsFile:  config.GlobalSettingsFile,
+			M2Path:              config.M2Path,
+		}
+		coordinates, err := utils.GetArtifactCoordinates(config.BuildTool, config.BuildDescriptorFile, options)
 		if err != nil {
 			return fmt.Errorf("failed to get build artifact description: %w", err)
 		}
 
 		nameTmpl := `{{list .GroupID .ArtifactID | join "-" | trimAll "-"}}`
 		name, version := versioning.DetermineProjectCoordinates(nameTmpl, config.VersioningModel, coordinates)
-		if config.ProjectName == "" {
+		if scan.aggregateProjectName == "" {
 			log.Entry().Infof("Resolved project name '%s' from descriptor file", name)
-			config.ProjectName = name
+			scan.aggregateProjectName = name
 		}
 		if config.ProductVersion == "" {
 			log.Entry().Infof("Resolved product version '%s' from descriptor file with versioning '%s'",
 				version, config.VersioningModel)
 			config.ProductVersion = version
 		}
 	}
+	scan.projectVersion = config.ProductVersion
```

Comment:
Project version is always equal to product version?

Reply:
Yes.

Reply:
Renamed it to scan.productVersion to make it consistent and clear.